File: types/type.go
package types

import (
	"regexp"
	"sort"

	"github.com/attic-labs/noms/d"
	"github.com/attic-labs/noms/hash"
)

// Type defines and describes Noms types, both custom and built-in.
// Desc provides more details of the type. It may contain only a types.NomsKind, in the case of
//     primitives, or it may contain additional information -- e.g. element Types for compound type
//     specializations, field descriptions for structs, etc. Either way, checking Kind() allows code
//     to understand how to interpret the rest of the data.
// If Kind() refers to a primitive, then Desc has no more info.
// If Kind() refers to List, Map, Set or Ref, then Desc is a list of Types describing the element type(s).
// If Kind() refers to Struct, then Desc contains a []Field.

type Type struct {
	Desc TypeDesc
	h    *hash.Hash
}

var typeForType = makePrimitiveType(TypeKind)

// Describe generate text that should parse into the struct being described.
func (t *Type) Describe() (out string) {
	return EncodedValue(t)
}

func (t *Type) Kind() NomsKind {
	return t.Desc.Kind()
}

func (t *Type) Name() string {
	// TODO: Remove from Type
	d.Chk.IsType(StructDesc{}, t.Desc, "Name only works on Struct types")
	return t.Desc.(StructDesc).Name
}

// Value interface
func (t *Type) Equals(other Value) (res bool) {
	return other != nil && t.Hash() == other.Hash()
}

func (t *Type) Less(other Value) (res bool) {
	return valueLess(t, other)
}

func (t *Type) Hash() hash.Hash {
	return EnsureHash(t.h, t)
}

func (t *Type) ChildValues() (res []Value) {
	switch desc := t.Desc.(type) {
	case CompoundDesc:
		for _, t := range desc.ElemTypes {
			res = append(res, t)
		}
	case StructDesc:
		desc.IterFields(func(name string, t *Type) {
			res = append(res, t)
		})
	case PrimitiveDesc:
		// Nothing, these have no child values
	default:
		d.Chk.Fail("Unexpected type desc implementation: %#v", t)
	}
	return
}

func (t *Type) Chunks() (chunks []Ref) {
	return
}

func (t *Type) Type() *Type {
	return typeForType
}

func MakePrimitiveType(k NomsKind) *Type {
	switch k {
	case BoolKind:
		return BoolType
	case NumberKind:
		return NumberType
	case StringKind:
		return StringType
	case BlobKind:
		return BlobType
	case ValueKind:
		return ValueType
	case TypeKind:
		return TypeType
	}
	d.Chk.Fail("invalid NomsKind: %d", k)
	return nil
}

func makePrimitiveType(k NomsKind) *Type {
	return buildType(PrimitiveDesc(k))
}

func MakePrimitiveTypeByString(p string) *Type {
	switch p {
	case "Bool":
		return BoolType
	case "Number":
		return NumberType
	case "String":
		return StringType
	case "Blob":
		return BlobType
	case "Value":
		return ValueType
	case "Type":
		return TypeType
	}
	d.Chk.Fail("invalid type string: %s", p)
	return nil
}

func MakeStructType(name string, fields map[string]*Type) *Type {
	for fn := range fields {
		verifyFieldName(fn)
	}
	return buildType(StructDesc{name, fields})
}

var fieldNameRe = regexp.MustCompile(`^[a-zA-Z][a-zA-Z0-9_]*$`)

func verifyFieldName(name string) {
	d.Exp.True(fieldNameRe.MatchString(name), "Invalid struct field name: %s", name)
}

func MakeListType(elemType *Type) *Type {
	return buildType(CompoundDesc{ListKind, []*Type{elemType}})
}

func MakeSetType(elemType *Type) *Type {
	return buildType(CompoundDesc{SetKind, []*Type{elemType}})
}

func MakeMapType(keyType, valType *Type) *Type {
	return buildType(CompoundDesc{MapKind, []*Type{keyType, valType}})
}

func MakeRefType(elemType *Type) *Type {
	return buildType(CompoundDesc{RefKind, []*Type{elemType}})
}

type unionTypes []*Type

func (uts unionTypes) Len() int           { return len(uts) }
func (uts unionTypes) Less(i, j int) bool { return uts[i].Hash().Less(uts[j].Hash()) }
func (uts unionTypes) Swap(i, j int)      { uts[i], uts[j] = uts[j], uts[i] }

// MakeUnionType creates a new union type unless the elemTypes can be folded into a single non union type.
func MakeUnionType(elemTypes ...*Type) *Type {
	seenTypes := map[hash.Hash]bool{}
	ts := flattenUnionTypes(elemTypes, &seenTypes)
	if len(ts) == 1 {
		return ts[0]
	}
	sort.Sort(unionTypes(ts))
	return buildType(CompoundDesc{UnionKind, ts})
}

func flattenUnionTypes(ts []*Type, seenTypes *map[hash.Hash]bool) []*Type {
	if len(ts) == 0 {
		return ts
	}

	ts2 := make([]*Type, 0, len(ts))
	for _, t := range ts {
		if t.Kind() == UnionKind {
			ts2 = append(ts2, flattenUnionTypes(t.Desc.(CompoundDesc).ElemTypes, seenTypes)...)
		} else {
			if !(*seenTypes)[t.Hash()] {
				(*seenTypes)[t.Hash()] = true
				ts2 = append(ts2, t)
			}
		}
	}
	return ts2
}

func buildType(desc TypeDesc) *Type {
	return &Type{Desc: desc, h: &hash.Hash{}}
}

var NumberType = makePrimitiveType(NumberKind)
var BoolType = makePrimitiveType(BoolKind)
var StringType = makePrimitiveType(StringKind)
var BlobType = makePrimitiveType(BlobKind)
var TypeType = makePrimitiveType(TypeKind)
var ValueType = makePrimitiveType(ValueKind)

File: js/src/type.js
// @flow

// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

import type Hash from './hash.js';
import Ref from './ref.js';
import type {NomsKind} from './noms-kind.js';
import {invariant} from './assert.js';
import {isPrimitiveKind, Kind} from './noms-kind.js';
import {ValueBase} from './value.js';
import type Value from './value.js';
import {compare, equals} from './compare.js';
import {describeType} from './encode-human-readable.js';
import search from './binary-search.js';

export interface TypeDesc {
  kind: NomsKind;
  equals(other: TypeDesc): boolean;
}

export class PrimitiveDesc {
  kind: NomsKind;

  constructor(kind: NomsKind) {
    this.kind = kind;
  }

  equals(other: TypeDesc): boolean {
    return other instanceof PrimitiveDesc && other.kind === this.kind;
  }
}

export class CompoundDesc {
  kind: NomsKind;
  elemTypes: Array<Type>;

  constructor(kind: NomsKind, elemTypes: Array<Type>) {
    this.kind = kind;
    this.elemTypes = elemTypes;
  }


  equals(other: TypeDesc): boolean {
    if (other instanceof CompoundDesc) {
      if (this.kind !== other.kind || this.elemTypes.length !== other.elemTypes.length) {
        return false;
      }

      for (let i = 0; i < this.elemTypes.length; i++) {
        if (!equals(this.elemTypes[i], other.elemTypes[i])) {
          return false;
        }
      }

      return true;
    }

    return false;
  }
}

export type Field = {
  name: string;
  type: Type;
};

export class StructDesc {
  name: string;
  fields: Field[];

  constructor(name: string, fields: Field[]) {
    this.name = name;
    this.fields = fields;
  }

  get fieldCount(): number {
    return this.fields.length;
  }

  get kind(): NomsKind {
    return Kind.Struct;
  }

  equals(other: TypeDesc): boolean {
    if (this === other) {
      return true;
    }

    if (other.kind !== this.kind) {
      return false;
    }
    invariant(other instanceof StructDesc);

    const fields = this.fields;
    const otherFields = other.fields;

    if (fields.length !== otherFields.length) {
      return false;
    }

    for (let i = 0; i < fields.length; i++) {
      if (fields[i].name !== otherFields[i].name || !equals(fields[i].type, otherFields[i].type)) {
        return false;
      }
    }

    return true;
  }

  forEachField(cb: (name: string, type: Type) => void) {
    const fields = this.fields;
    for (let i = 0; i < fields.length; i++) {
      cb(fields[i].name, fields[i].type);
    }
  }

  getField(name: string): ?Type {
    const f = findField(name, this.fields);
    return f && f.type;
  }

  setField(name: string, type: Type) {
    const f = findField(name, this.fields);
    if (!f) {
      throw new Error(`No such field "${name}"`);
    }
    f.type = type;
  }
}

function findField(name: string, fields: Field[]): ?Field {
  const i = findFieldIndex(name, fields);
  return i !== -1 ? fields[i] : undefined;
}

/**
 * Finds the index of the `Field` or `-1` if not found.
 */
export function findFieldIndex(name: string, fields: Field[]): number {
  const i = search(fields.length, i => {
    const n = fields[i].name;
    return n === name ? 0 : n > name ? 1 : -1;
  });
  return i === fields.length || fields[i].name !== name ? -1 : i;
}


export class Type<T: TypeDesc> extends ValueBase {
  _desc: T;

  constructor(desc: T) {
    super();
    this._desc = desc;
  }

  get type(): Type {
    return typeType;
  }

  get chunks(): Array<Ref> {
    return [];
  }

  get kind(): NomsKind {
    return this._desc.kind;
  }

  get desc(): T {
    return this._desc;
  }

  get name(): string {
    invariant(this._desc instanceof StructDesc);
    return this._desc.name;
  }

  get elemTypes(): Array<Type> {
    invariant(this._desc instanceof CompoundDesc);
    return this._desc.elemTypes;
  }

  describe(): string {
    return describeType(this);
  }
}

function buildType<T: TypeDesc>(desc: T): Type<T> {
  return new Type(desc);
}

function makePrimitiveType(k: NomsKind): Type<PrimitiveDesc> {
  return buildType(new PrimitiveDesc(k));
}

export function makeListType(elemType: Type): Type<CompoundDesc> {
  return buildType(new CompoundDesc(Kind.List, [elemType]));
}

export function makeSetType(elemType: Type): Type<CompoundDesc> {
  return buildType(new CompoundDesc(Kind.Set, [elemType]));
}

export function makeMapType(keyType: Type, valueType: Type): Type<CompoundDesc> {
  return buildType(new CompoundDesc(Kind.Map, [keyType, valueType]));
}

export function makeRefType(elemType: Type): Type<CompoundDesc> {
  return buildType(new CompoundDesc(Kind.Ref, [elemType]));
}

export function makeStructType(name: string, fields: {[key: string]: Type}): Type<StructDesc> {
  verifyStructName(name);
  const keys = Object.keys(fields);
  keys.sort();
  const fs = new Array(keys.length);
  for (let i = 0; i < keys.length; i++) {
    const name = keys[i];
    verifyFieldName(name);
    const type = fields[name];
    fs[i] = {name, type};
  }

  return buildType(new StructDesc(name, fs));
}

/**
 * makeUnionType creates a new union type unless the elemTypes can be folded into a single non
 * union type.
 */
export function makeUnionType(types: Type[]): Type {
  types = flattenUnionTypes(types, Object.create(null));
  if (types.length === 1) {
    return types[0];
  }
  types.sort(compare);
  return buildType(new CompoundDesc(Kind.Union, types));
}

const fieldNameRe = /^[a-zA-Z][a-zA-Z0-9_]*$/;

function verifyName(name: string, kind: '' | ' field') {
  if (!fieldNameRe.test(name)) {
    throw new Error(`Invalid struct${kind} name: "${name}"`);
  }
}

function verifyFieldName(name: string) {
  verifyName(name, ' field');
}

function verifyStructName(name: string) {
  if (name !== '') {
    verifyName(name, '');
  }
}

function flattenUnionTypes(types: Type[], seenTypes: {[key: Hash]: boolean}): Type[] {
  if (types.length === 0) {
    return types;
  }

  const newTypes = [];
  for (let i = 0; i < types.length; i++) {
    if (types[i].kind === Kind.Union) {
      newTypes.push(...flattenUnionTypes(types[i].desc.elemTypes, seenTypes));
    } else {
      if (!seenTypes[types[i].hash]) {
        seenTypes[types[i].hash] = true;
        newTypes.push(types[i]);
      }
    }
  }
  return newTypes;
}

export const boolType = makePrimitiveType(Kind.Bool);
export const numberType = makePrimitiveType(Kind.Number);
export const stringType = makePrimitiveType(Kind.String);
export const blobType = makePrimitiveType(Kind.Blob);
export const typeType = makePrimitiveType(Kind.Type);
export const valueType = makePrimitiveType(Kind.Value);

export const refOfBlobType = makeRefType(blobType);
export const refOfValueType = makeRefType(valueType);
export const listOfValueType = makeListType(valueType);
export const setOfValueType = makeSetType(valueType);
export const mapOfValueType = makeMapType(valueType, valueType);

/**
 * Gives the existing primitive Type value for a NomsKind.
 */
export function getPrimitiveType(k: NomsKind): Type {
  invariant(isPrimitiveKind(k));
  switch (k) {
    case Kind.Bool:
      return boolType;
    case Kind.Number:
      return numberType;
    case Kind.String:
      return stringType;
    case Kind.Blob:
      return blobType;
    case Kind.Type:
      return typeType;
    case Kind.Value:
      return valueType;
    default:
      invariant(false, 'not reachable');
  }
}

// Returns the Noms type of any value. This will throw if you pass in an object that cannot be
// represented by noms.
export function getTypeOfValue(v: Value): Type {
  if (v instanceof ValueBase) {
    return v.type;
  }

  switch (typeof v) {
    case 'string':
      return stringType;
    case 'boolean':
      return boolType;
    case 'number':
      return numberType;
    default:
      throw new Error('Unknown type');
  }
}
